charm/main.py

```diff
@@ -0,0 +1,218 @@
+
+
+class BaseCommand:
+
+    name = None
+    help_msg = None
+
+    def __init__(self, group):
+        if self.name is None or self.help_msg is None:
+            raise RuntimeError("Command not properly created: {!r}".format(self.__class__))
+        self.group = group
+
+    def fill_parser(self, parser):
+        """Overwrite in each command and fill the parser with command-specific parameters.
+
+        If not overwritten, the command will not have any parameters.
+        """
+
+
+class VersionCommand(BaseCommand):
+    """Show the version."""
+    name = 'version'
+    help_msg = "show the version"
+
+    def run(self, parsed_args):
+        """Run the command."""
+        # XXX: we need to define how we want to store the project version (in config/file/etc.)
+        version = '0.0.1'
+        logger.info('Version: %s', version)
+
+
+# XXX: while VersionCommand above probably would stay here, the other commands surely will
+# be located in different files... that said, I'm keeping these examples here, for
+# simplicitiy
+
+
+class CommandExampleDebug(BaseCommand):
+    """Just an example."""
+
+    name = 'example-debug'
+    help_msg = "show msg in debug"
+
+    def fill_parser(self, parser):
+        parser.add_argument('foo')
+        parser.add_argument('--bar', action='store_true', help="To use this command in a bar")
+
+    def run(self, parsed_args):
+        logger.debug(
+            "Example showing log in DEBUG: foo=%s bar=%s", parsed_args.foo, parsed_args.bar)
+
+
+class CommandExampleError(BaseCommand):
+    """Just an example."""
+
+    name = 'example-error'
+    help_msg = "show msg in error"
+
+    def run(self, parsed_args):
+        logger.error("Example showing log in ERROR.")
+
+
+# Collect commands in different groups, for easier human consumption. Note that this is not
+# declared in each command because it's much easier to do this separation/grouping in one
+# central place and not distributed in several classes/files.
+COMMANDS_GROUPS = [
+    ('basic', [VersionCommand, CommandExampleError]),
+    ('advanced', [CommandExampleDebug]),
+]
+
+
+class CustomArgumentParser(argparse.ArgumentParser):
+    """ArgumentParser with grouped commands help."""
+    # XXX: we should find a better way of doing this
+
+    def format_help(self):
+        """SUPER HACKY help."""
+        main = False
+        for ag in self._action_groups:
+            if ag.title == 'MARKER':
+                self._action_groups.remove(ag)
+                main = True
+                break
+        base = super().format_help()
+        if not main:
+            return base
+
+        extra = ['', 'commands:', '']
+        for group, cmd_classes in COMMANDS_GROUPS:
+            extra.append("    title for {!r}:".format(group))
+            for cmd_class in cmd_classes:
+                extra.append("        {:20s} {}".format(cmd_class.name, cmd_class.help_msg))
+            extra.append('')
+
+        return base + '\n'.join(extra)
```

Comment:
you say this is hacky, but I think it's fine

Reply:
Well, we're accesing internal private atributes of ArgumentParser, so it may break in the future.